Update Jetpack Compose to dev15

wardellbagby opened the pull request:
Split along 3 commits:

1. Updates the dependency versions, artifact coordinates, and removes some now unnecessary stuff with Kotlin 1.4 being used.
2. Updates our shippable lib code to use the new dev15 APIs.
3. Fixes the UI tests.

Most of this is fairly rote work, but there are three things a bit weird here.

1. Semantics merging isn't happening as it was back in dev12. Issue is here https://issuetracker.google.com/issues/161979921
2. Can't compile the "use regular views in Compose" syntax, a la `HostView(viewFactory = viewFactory, update = Pair(rendering, wrappedEnvironment))`
3. `OwnerAmbient` is gone now, so I replaced its usages with `ViewAmbient`.

1 worries me a lot; I'm worried it could've have unintentional consequences with how I resolved it.
2 worries me a little; I'm worried that that shorthand could've created the Views a different way than I did.
3 worries me a little; I'm worried that `ViewAmbient` is returning a view maybe not suitable to be a parent? But I am unsure.

For ease of reviewing, here's the entire commit message for the 3 commits in this PR.


Update Compose to dev15 - Dependencies
- Updates the Compose version to dev15.
- Updates the coordinates for various Compose libs that have changed.
- Updates Kotlin to 1.4-M3, since that's required by Compose now.
- Removes the code forcing the Kotlin compiler to use API version 1.3.
- Updates the README to reflect these changes.


Update Compose to dev15 - API Changes - Production Code
- Removes ComposeSupport.kt and its associated tests, replacing its usages with the
setContent method that now exists in the Compose API.
- Updates various Compose API usages that were removed or replaced.
   - mutableStateOf and its areEquivalent parameter have been replaced
by a SnapshotMutationPolicy
  - Removes Modifier.ripple usages entirely, since that is now the default animations
for clicks.
  - Updates toPx usages to no longer call .value, since those now return a Float
directly instead of a Px object.
  - Opts into the usage of FlowRow, since that's now experimental.
  - Scaffold renamed topAppBar to topBar.
  - OwnerAmbient is removed entirely. Instead, we now use the ViewAmbient to get a
valid parent view.
  - Our usage of our internal HostView no longer compiled, so manually creating it
using the current Context.


Update Compose to dev15 - API Changes - UI Tests
- Updates various Compose API usages that were removed or replaced.
   - findByText => onNodeWithText
  - findAllByText => onAllNodesWithText
  - runOnIdleCompose => runOnIdle
  - doClick => performClick
  - FrameManager.framed is now unneeded.
  - Similar to HostView, RootView no longer compiled, so calls it like a normal
View but wraps it in the AndroidView composable, so that its added to the
composition as expected.
- Adds new UI tests for testing rendering legacy Android views when in Compose.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/square/workflow-kotlin-compose?pullRequest=56) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/square/workflow-kotlin-compose?pullRequest=56) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/square/workflow-kotlin-compose?pullRequest=56) it.</sub>

zach-klippenstein commented on README.md:
````diff
@@ -48,20 +48,12 @@ android {
 }
 ```
 
-You may also need to set the Kotlin API version to 1.3:
-
-```groovy
-tasks.withType(org.jetbrains.kotlin.gradle.tasks.KotlinCompile).configureEach {
-  kotlinOptions.apiVersion = "1.3"
-}
-```
````
🎉

zach-klippenstein commented on build.gradle.kts:
```diff
@@ -33,8 +33,9 @@ buildscript {
     mavenCentral()
     gradlePluginPortal()
     google()
+    maven("https://dl.bintray.com/kotlin/kotlin-eap")
```
```suggestion
    // For Kotlin 1.4.
    maven("https://dl.bintray.com/kotlin/kotlin-eap")
```

zach-klippenstein commented on build.gradle.kts:
```diff
@@ -46,6 +47,7 @@ subprojects {
     google()
     mavenCentral()
     jcenter()
+    maven("https://dl.bintray.com/kotlin/kotlin-eap")
```
```suggestion
    // For Kotlin 1.4.
    maven("https://dl.bintray.com/kotlin/kotlin-eap")
```

zach-klippenstein commented on build.gradle.kts:
```diff
@@ -84,8 +86,12 @@ subprojects {
+      // to define our own experiments, and some required args for compose dev15 taken from
+      // https://developer.android.com/jetpack/androidx/releases/compose-runtime
+      freeCompilerArgs += listOf(
+          "-Xopt-in=kotlin.RequiresOptIn", "-Xallow-jvm-ir-dependencies", "-Xskip-prerelease-check"
```
```suggestion
          "-Xopt-in=kotlin.RequiresOptIn",
          "-Xallow-jvm-ir-dependencies",
          "-Xskip-prerelease-check"
```

zach-klippenstein commented on buildSrc/src/main/java/Dependencies.kt:
```diff
@@ -19,25 +19,26 @@ import java.util.Locale.US
 import kotlin.reflect.full.declaredMembers
 
 object Versions {
-  const val compose = "0.1.0-dev12"
```
I kind of want to do an update to dev14 first, since this is a pretty major update, and I'd like to be able to track what changed on each release. But I don't want to block this release any more, so probably won't.

zach-klippenstein commented on buildSrc/src/main/java/Dependencies.kt:
```diff
@@ -19,25 +19,26 @@ import java.util.Locale.US
-  const val compose = "0.1.0-dev12"
-  const val kotlin = "1.3.71"
+  const val compose = "0.1.0-dev15"
+  const val kotlin = "1.4-M3"
```
🙌

zach-klippenstein commented on compose-tooling/build.gradle.kts:
```diff
@@ -30,9 +30,6 @@ apply(from = rootProject.file(".buildscript/configure-android-defaults.gradle"))
 apply(from = rootProject.file(".buildscript/android-ui-tests.gradle"))
 
 apply(from = rootProject.file(".buildscript/configure-compose.gradle"))
-tasks.withType<KotlinCompile> {
-  kotlinOptions.apiVersion = "1.3"
-}
```
🎉

zach-klippenstein commented on core-compose/src/main/java/com/squareup/workflow/ui/compose/ComposeViewFactory.kt:
```diff
@@ -110,12 +112,10 @@ internal class ComposeViewFactory<RenderingT : Any>(
           initialRendering,
           initialViewEnvironment
       ) { rendering, environment ->
-        FrameManager.framed {
```
🙌

zach-klippenstein commented on core-compose/src/main/java/com/squareup/workflow/ui/compose/ComposeViewFactory.kt:
```diff
@@ -131,7 +131,7 @@ internal class ComposeViewFactory<RenderingT : Any>(
       ) { rendering, environment ->
         // Entry point to the world of Compose.
-        composeContainer.setContent(parentComposition) {
+        composeContainer.setContent(Recomposer.current(), parentComposition) {
```
😍

zach-klippenstein commented on core-compose/src/main/java/com/squareup/workflow/ui/compose/internal/ViewFactories.kt:
```diff
@@ -93,8 +98,15 @@ import kotlin.properties.Delegates.observable
   val wrappedEnvironment = remember(viewEnvironment) {
     viewEnvironment + (ParentComposition to parentComposition)
   }
-
-  HostView(viewFactory = viewFactory, update = Pair(rendering, wrappedEnvironment))
+  val context = ContextAmbient.current
+  val view = remember(context) { HostView(context) }.apply {
+    this.viewFactory = viewFactory
+    update = Pair(rendering, wrappedEnvironment)
+  }
+  AndroidView(
+      view = view,
+      modifier = Modifier.testTag(LegacyAndroidViewTestTag)
+  )
```
I think we can use the new `emitView` API here instead. Gonna try it locally and report back.

zach-klippenstein replied:
Yep, it works! Let me know if you want me to post the code I ended up with.

zach-klippenstein commented on samples/src/main/java/com/squareup/sample/hellocompose/HelloBinding.kt:
```diff
@@ -20,15 +20,13 @@ import androidx.ui.core.Modifier
   Text(
       rendering.message,
       modifier = Modifier
-          .ripple()
```
🔥

zach-klippenstein commented on samples/src/main/java/com/squareup/sample/launcher/SampleLauncherApp.kt:
```diff
@@ -75,9 +74,7 @@ import com.squareup.sample.R
-  // See https://issuetracker.google.com/issues/156875705.
-  @Suppress("DEPRECATION")
-  val rootView = (OwnerAmbient.current as AndroidOwner).view
+  val rootView = ViewAmbient.current
```
🎉

zach-klippenstein commented on compose-tooling/src/androidTest/java/com/squareup/workflow/ui/compose/tooling/PreviewComposeWorkflowTest.kt:
```diff
@@ -81,25 +80,23 @@ class PreviewComposeWorkflowTest {
-        Semantics(container = true, mergeAllDescendants = true) {
-          WorkflowRendering(props.second, environment)
-        }
+        WorkflowRendering(props.second, environment)
```
Seems good that this is the new default behavior.

zach-klippenstein commented on core-compose/src/androidTest/java/com/squareup/workflow/ui/compose/ComposeViewFactoryTest.kt:
```diff
@@ -47,18 +48,18 @@ class ComposeViewFactoryTest {
-      // This is valid Compose code, but the IDE doesn't know that yet so it will show an
-      // unsuppressable error.
-      RootView(viewEnvironment = viewEnvironment)
+      AndroidView(RootView(ContextAmbient.current).apply { setViewEnvironment(viewEnvironment) })
```
Why doesn't the view need to be remembered here, like in the previous commit?

wardellbagby replied:
It actually never _needs_ to be remembered, but that keeps the view from being unnecessarily recreated every time we rerender the composition. 

That's unnecessary here in the UI test, _but_, full disclosure, I did the `remember` _after_ I had already changed this test. For consistency, I can change it, but it doesn't really matter, I don't think.

zach-klippenstein commented on core-compose/src/androidTest/java/com/squareup/workflow/ui/compose/internal/ViewFactoriesTest.kt:
```diff
@@ -48,14 +60,91 @@ class ViewFactoriesTest {
+        .componentNode
+        .owner!! as AndroidOwner
+
+    // Really digging into internals here. This would work better if the accessibility
+    // inter-op when creating a LayoutNode from a View existed, so that we could continue
+    // to use the standard Compose testing API when dealing with Android views.
```
Did you file a Compose issue for this? https://issuetracker.google.com/issues?q=componentid:612128

wardellbagby replied:
I didn't. There's a TODO code comment where the update would happen at (`AndroidViewHolder.toLayoutNode()`) that would fix this. Think it's still worth filing?

zach-klippenstein commented on core-compose/src/androidTest/java/com/squareup/workflow/ui/compose/internal/ViewFactoriesTest.kt:
```diff
@@ -48,14 +60,91 @@ class ViewFactoriesTest {
+    onNodeWithText("two", useUnmergedTree = true).assertIsDisplayed()
+  }
+
+  @Test fun WorkflowRendering_rendersLegacyAndroidView() {
```
These are great, thanks for adding these tests!

zach-klippenstein commented:
> 3 worries me a little; I'm worried that ViewAmbient is returning a view maybe not suitable to be a parent? But I am unsure.

I'm pretty sure `ViewAmbient` is the API we actually wanted here from the start, per that issue that was linked here.

zach-klippenstein commented on compose-tooling/build.gradle.kts:
```diff
@@ -1,4 +1,4 @@
-import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+
```
Extra line.
```suggestion
```

zach-klippenstein commented on core-compose/src/androidTest/java/com/squareup/workflow/ui/compose/CompositionRootTest.kt:
```diff
@@ -49,7 +47,10 @@ class CompositionRootTest {
-    findByText("one\ntwo").assertIsDisplayed()
+    // These semantics used to merge, but as of dev15, they don't, which seems to be a bug.
+    // https://issuetracker.google.com/issues/161979921
+    onNodeWithText("one", useUnmergedTree = true).assertIsDisplayed()
+    onNodeWithText("two", useUnmergedTree = true).assertIsDisplayed()
```
I tried removing this `useUnmergedTree` argument locally, and the test still passed.

zach-klippenstein commented on core-compose/src/androidTest/java/com/squareup/workflow/ui/compose/internal/ViewFactoriesTest.kt:
```diff
@@ -48,14 +60,91 @@ class ViewFactoriesTest {
+
+  @Test fun WorkflowRendering_rendersLegacyAndroidView() {
+    val viewEnvironment = ViewEnvironment(ViewRegistry(LegacyViewViewFactory))
+        .withCompositionRoot { content ->
+          Column {
+            Text("one")
+            content()
+          }
```
This test doesn't need to assert anything about the composition root, it's unrelated and just makes this test more confusing.

zach-klippenstein commented on core-compose/src/androidTest/java/com/squareup/workflow/ui/compose/internal/ViewFactoriesTest.kt:
```diff
@@ -48,14 +60,91 @@ class ViewFactoriesTest {
+    }
+
+    onNodeWithText("one", useUnmergedTree = true).assertDoesNotExist()
+    onNodeWithTag(LegacyAndroidViewTestTag, useUnmergedTree = true).assertIsDisplayed()
```
Instead of using a test tag here, let's add a dependency on Espresso (to `android-ui-tests.gradle`), and then actually assert what we need to here: `onView(withText("two")).check(matches(isDisplayed()))`. I just verified this locally, and it works (also don't need the `useUnmergedTree` arguments). This makes the tests better (actually testing that the view we care about shown correctly), more readable, and means we don't have to wrap the `emitView` call with something just to pass a test modifier.

zach-klippenstein commented:
Left a bunch of comments, feel free to address as separate commits – I'll ask you to squash all of these into a single commit before merging anyway, so we don't end up with unbuildable commits in the history. Thanks again, this is fantastic!

wardellbagby commented:
I believe I've addressed all the comments now, except for 1:

Can't change `SampleLauncherTest#allSamplesLaunch` to not use `useUnmergedTree`; the description text just isn't in the semantics tree at all.

wardellbagby commented:
Oh, also pre-emptively squashed the whole thing down.

zach-klippenstein commented on core-compose/src/androidTest/java/com/squareup/workflow/ui/compose/ComposeViewFactoryTest.kt:
```diff
@@ -47,18 +47,21 @@ class ComposeViewFactoryTest {
-      // This is valid Compose code, but the IDE doesn't know that yet so it will show an
-      // unsuppressable error.
-      RootView(viewEnvironment = viewEnvironment)
+      emitView(
+          ctor = ::RootView,
+          update = { it.setViewEnvironment(viewEnvironment) }
+      )
```
Nit: More idiomatic to write like this:
```suggestion
      emitView(::RootView) {
        it.setViewEnvironment(viewEnvironment)
      }
```

zach-klippenstein commented on core-compose/src/main/java/com/squareup/workflow/ui/compose/internal/ViewFactories.kt:
```diff
@@ -94,7 +95,12 @@ import kotlin.properties.Delegates.observable
   }
 
-  HostView(viewFactory = viewFactory, update = Pair(rendering, wrappedEnvironment))
+  emitView(
+      ctor = ::HostView,
+      update = {
+        it.viewFactory = viewFactory
+        it.update = Pair(rendering, wrappedEnvironment)
+      })
```
```suggestion
  emitView(::HostView) {
    it.viewFactory = viewFactory
    it.update = Pair(rendering, wrappedEnvironment)
  }
```
